runtime/src/main/scala/com/armanbilge/vilcacora/runtime/Interpreter.scala

```diff
@@ -0,0 +1,305 @@
+
+    val inputA = memory(op.inputs(0)).asInstanceOf[Ptr[CFloat]]
+    val inputB = memory(op.inputs(1)).asInstanceOf[Ptr[CFloat]]
+    val output = memory(op.outputs(0)).asInstanceOf[Ptr[CFloat]]
+
+    for (i <- 0 until count)
+      !(output + i) = f(!(inputA + i), !(inputB + i))
+  }
+
+  private def handleCast(op: Operation.Cast, memory: MemoryMap, model: ModelIR): IO[Unit] = IO {
+    val inputAlloc = model.allocations(op.input)
+    val outputAlloc = model.allocations(op.output)
+
+    val inputPtr = memory(op.input)
+    val outputPtr = memory(op.output)
+    val count = inputAlloc.shape.product
+
+    (inputAlloc.dataType, outputAlloc.dataType) match {
+      case (from, to) if from == to =>
+        val _ = memcpy(outputPtr, inputPtr, (count * from.sizeInBytes).toUSize)
+      case (DataType.Float64, DataType.Float32) =>
+        val in = inputPtr.asInstanceOf[Ptr[CDouble]]
+        val out = outputPtr.asInstanceOf[Ptr[CFloat]]
+        for (i <- 0 until count)
+          !(out + i) = (!(in + i)).toFloat
+      case (DataType.Float32, DataType.Float64) =>
+        val in = inputPtr.asInstanceOf[Ptr[CFloat]]
+        val out = outputPtr.asInstanceOf[Ptr[CDouble]]
+        for (i <- 0 until count)
+          !(out + i) = (!(in + i)).toDouble
+      // Add more cases for other data types
+      case (from, to) =>
+        throw new NotImplementedError(s"Cast from $from to $to is not implemented.")
+    }
+  }
+
+  private def handleSvmClassifier(
+      op: Operation.SVMClassifier,
+      memory: MemoryMap,
+      model: ModelIR,
+  ): IO[Unit] = {
+    // The number of features is the size of the last dimension of the input tensor.
+    val numFeatures = model.allocations(op.input).shape.last
+
+    val svmResources = for {
+      modelPtr <- buildSvmModelResource(op, numFeatures)
+      // Allocate space for features + 1 for the terminator node.
+      svmInputNode <- malloc[svm_node](sizeof[svm_node] * (numFeatures + 1).toUSize)
+    } yield (modelPtr, svmInputNode)
+
+    svmResources.use { case (modelPtr, svmInputNode) =>
+      // Using a for-comprehension here breaks the logic into clean, sequential IO steps.
+      for {
+        inputPtr <- IO.pure(memory(op.input).asInstanceOf[Ptr[CDouble]])
+
+        // This entire block is now a single IO action that performs the C interop.
+        predictionResult <- IO {
+          // 1. Fill the svm_node struct with input data from the input tensor.
+          for (i <- 0 until numFeatures) {
+            val node = svmInputNode + i
+            node.index = i + 1 // LIBSVM indices are 1-based.
+            node.value = !(inputPtr + i)
+          }
+          (svmInputNode + numFeatures).index = -1 // Terminator node.
+
+          // 2. Prepare for and call the prediction function.
+          val nrClass = op.classLabels.size
+          val decValuesCount = nrClass * (nrClass - 1) / 2
+          val decisionValuesPtr = stackalloc[CDouble](decValuesCount.toUInt)
```

Comment:
I see that you are returning `decisionValuesPtr` at the end of the `IO`, but this is not safe to do with `stackalloc`ated memory. The stack is popped at the end of the `IO`, so that memory is no longer safe to access.